### ISSUE TITLE
JuMP return code mishandling leading to inconsistency

### ISSUE THREAD

**Issue by tomerarnon** [sisl/NeuralVerification.jl#71]
I think the cause of #70 is that we're mishandling some `JuMP` return codes. We test for `OPTIMAL` and `INFEASIBLE`, but we are sometimes also seeing `INFEASIBLE_OR_UNBOUNDED`. 

First we should consider whether any of the objectives we use can ever by unbounded in the first place. If not, we can always treat the problems as infeasible in these cases. If it _is_ possible, then we need to consider what to do in unbounded cases. E.g., consider MIPVerify:

```julia
##########################################
# currently
o = objective_value(model)
if termination_status(model) == INFEASIBLE
    return AdversarialResult(:holds)
end
if o >= maximum(problem.input.radius)
    return AdversarialResult(:holds)
else
    return AdversarialResult(:violated, o)
end
##########################################


# fixed?
o = objective_value(model)
if isfeasible(model) && o < maximum(problem.input.radius)
    return AdversarialResult(:violated, o)
else
    return AdversarialResult(:holds, o)
end
```
(Note that `isfeasible` is some custom function that tests for feasibility according to our criteria, solving a feasibility problem if necessary).

In the case where `model` is feasible but unbounded, we would never want to still return `o` as an adversarial result, given that it's NaN. 
In this particular case we're safe from that, since `NaN < some_num == false`, but in other solvers this may not be as cut and dry.

What about DLV:
```julia
# formerly: if termination_status(model) == OPTIMAL
if isfeasible(model)
    return (true, value(first(neurons)))
else
    return (false, [])
end
```
In the unbounded (but feasible) case, can we really still return `value(first(neurons)))` as the counterexample (it will probably be full of NaNs)? If not, given that the problem is indeed violated, what can we return instead?

Again, not sure if the problems we're dealing with can ever _actually_ be unbounded or whether that return code always applies to infeasibility. @changliuliu, I could definitely use your input on this! I'll be opening a PR shortly that begins to address this. Hopefully we can integrate it before next week.

**Comment by tomerarnon** [comment]
Upon second review, here's what I've come up with:

DLV, iLP, MIPVerify use `max_disturbance!`, which minimizes the infinity norm, and is therefore naturally bounded. MIPVerify, which _was_ affected by this bug, was affected because it was explicitly testing `INFEASIBLE`. If it had explicitly tested `OPTIMAL` instead, it would have been safe. This is a simple change to make in these three cases.

Reluplex and NSVerify solve feasibility problems, so should be simple to as well.

Planet, BaB, and Sherlock use potentially unbounded objectives, so need more careful consideration.

I'm not certain regarding Duality but I think it can be unbounded in some cases due to `activation_value` (I'm not sure since the `symbolic_max` might implicitly enforce positivity(?) which would make the minimization constraint bounded).

**Comment by changliuliu** [comment]
In those unbounded cases, is it always positive NaN? Is it possible to have negative NaN? If the signs of the NaN is correct, we may still be able to use those unbounded results, i.e., to check inequality constraints.

In addition, one way to get around of this boundness problem is to always require the input space to be bounded (or manually set a big enough bound).

**Comment by tomerarnon** [comment]
We can't get information from the NaN itself since `-NaN ==  NaN`, but we should always be able to know whether a particular problem is unbounded below or above by its definition (Min/Max). 

> If the signs of the NaN are correct, we may still be able to use those unbounded results, i.e., to check inequality constraints.

I'll try and think through how this will work.

For reference, examples of potentially unbounded objectives:

 Sherlock:
```julia
o = get_gradient(nnet, x) * first(neurons)
index = ifelse(type == :max, 1, -1)
@objective(model, Max, index * first(o)) # unbounded above
```

Bab:
```julia
index = ifelse(type == :max, 1, -1)
o = sum(last(neurons))
@objective(model, Max, index * o) # unbounded above
```

Planet is maybe the most complicated. Here is one instance from `tighten_bounds`
```julia
min_sum!(model, neurons) # unbounded below
optimize!(model)
# not an ok check, since a return code of INFEASIBLE_OR_UNBOUNDED can mean feasibility.
termination_status(model) == OPTIMAL || return (INFEASIBLE, bounds) 
```

If we figure out what to do in these three cases, we'll be in quite good shape on this. 

> In addition, one way to get around of this boundness problem is to always require the input space to be bounded (or manually set a big enough bound).

I think this is more of a last resort. It would be good not to artificially bound our halfspace constraints if we don't have to, since that would really throw off a lot of what we already do.

**Comment by changliuliu** [comment]
> It would be good not to artificially bound our halfspace constraints if we don't have to, since that would really throw off a lot of what we already do.

We only have halfspace constraints on the output side. And we only need to bound the input space for the three solvers you mentioned above to make the problem bounded. Actually, all three solvers take hyper-rectangle inputs which are already bounded. In theory, there shouldn't be any unbounded output. Did we really observe any failure case for these three solvers?

**Comment by tomerarnon** [comment]
> We only have halfspace constraints on the output side. And we only need to bound the input space for the three solvers you mentioned above to make the problem bounded. Actually, all three solvers take hyper-rectangle inputs which are already bounded.

Yes, I suppose that's true at the moment. We can definitely use this fact to patch this up. 
But we have discussed relaxing some restrictions as time goes on (do these algorithms require bounded inputs intrinsically or can that requirement be removed eventually?). 

>  In theory, there shouldn't be any unbounded output. Did we really observe any failure case for these three solvers?

Nope, we don't have a confirmed failure for these solvers, only noted that they're vulnerable. The only failure I'm certain of is Reluplex. Just taking the opportunity to remove the issue for good.

**Comment by mykelk** [comment]
Just a quick note of correction: `NaN != NaN` and `NaN != -NaN`.

**Comment by tomerarnon** [comment]
True, thanks for the correction!

**Comment by tomerarnon** [comment]
closed by a combination of #174 and #151
